torchrl/envs/libs/unity.py

```diff
@@ -0,0 +1,385 @@
+        for behavior_name in env.behavior_specs.keys():
+            decision_steps, terminal_steps = env.get_steps(behavior_name)
+            num_agents += len(decision_steps) + len(terminal_steps)
+        return num_agents
+
+    def _set_seed(self, seed: int | None):
+        warn(
+            "Seeding through _set_seed has not been implemented. Please set the "
+            "seed when you create the environment."
+        )
+
+    @_classproperty
+    def available_envs(cls) -> list[str]:
+        return []
+
+    def _build_env(self, env: BaseEnv):
+        if not env.behavior_specs:
+            # Take a single step so that the brain information will be sent over
+            env.step()
+        self._behavior_names = list(env.behavior_specs.keys())
+        self.num_agents = self._compute_num_agents(env)
+        self._agent_ids = torch.tensor(range(self.num_agents), dtype=torch.int)
+        self._agent_id_to_behavior_name = {}
+        return env
+
+    def _make_specs(self, env: BaseEnv) -> None:
+        observation_specs = [None] * self.num_agents
+        action_specs = [None] * self.num_agents
+        reward_specs = [None] * self.num_agents
+        done_specs = [None] * self.num_agents
+        valid_mask_specs = [None] * self.num_agents
+
+        for behavior_name, behavior_unity_spec in env.behavior_specs.items():
+            decision_steps, terminal_steps = env.get_steps(behavior_name)
+            for steps in [decision_steps, terminal_steps]:
+                for agent_id in steps.agent_id:
+                    self._agent_id_to_behavior_name[agent_id] = behavior_name
+
+                    observation_specs[agent_id] = CompositeSpec(
+                        {
+                            f"obs_{agent_id}_{i}": _unity_to_torchrl_spec_transform(
+                                spec, dtype=np.dtype("float32"), device=self.device
+                            )
+                            for i, spec in enumerate(
+                                behavior_unity_spec.observation_specs
+                            )
+                        }
+                    )
+                    action_specs[agent_id] = _unity_to_torchrl_spec_transform(
+                        behavior_unity_spec.action_spec,
+                        dtype=np.dtype("int32"),
+                        device=self.device,
+                    )
+                    reward_specs[agent_id] = UnboundedContinuousTensorSpec(
+                        shape=[1], device=self.device
+                    )
+                    done_specs[agent_id] = DiscreteTensorSpec(
+                        n=2, shape=[1], dtype=torch.bool, device=self.device
+                    )
+                    valid_mask_specs[agent_id] = DiscreteTensorSpec(
+                        n=2, shape=[1], dtype=torch.bool, device=self.device
+                    )
+
+        self.observation_spec = CompositeSpec(
+            {
+                "agents": CompositeSpec(
```

Comment:
What do you think about grouping agents by behaviour (as in case 2 of #1463)?

We would still have masks within behaviours for agent dropping in or out.

We could then allow users that do not want agents grouped by behaviour to change this by setting
`env = MLAgentsEnv(group_map={"group_name": [agent_names]}`

For example what you have now would be obtained by providing
`group_map={"agents",["agent_0",...,"agent_n]}`

but the default (whrn the map is not provided) would be groups by behavior